akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowStatefulMapConcatSpec.scala

```diff
@@ -0,0 +1,37 @@
+/**
+ * Copyright (C) 2016 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.scaladsl
+
+import akka.stream.ActorMaterializerSettings
+import akka.stream.testkit._
+
+class FlowStatefulMapConcatSpec extends AkkaSpec with ScriptedTest {
+
+  val settings = ActorMaterializerSettings(system)
+    .withInputBuffer(initialSize = 2, maxSize = 16)
+
+  "A StatefulMapConcat" must {
+
+    "work in happy case" in {
+      val script = Script(
+        Seq(2) -> Seq(),
+        Seq(1) -> Seq(1, 1),
+        Seq(3) -> Seq(3),
+        Seq(6) -> Seq(6, 6, 6))
+      TestConfig.RandomTestRange foreach (_ ⇒ runScript(script, settings)(_.statefulMapConcat(() ⇒ {
+        var prev: Option[Int] = None
+        x ⇒ prev match {
+          case Some(e) ⇒
+            prev = Some(x)
+            (1 to e) map (_ ⇒ x)
+          case None ⇒
+            prev = Some(x)
+            List.empty[Int]
+        }
+      })))
+    }
```

Comment:
Yes, this is how it is supposed to be used. Please add Java tests as well to make sure that type inference and syntax work out. One issue there is that Java only allows to close over effectively final references, which means that people will have to put their state into mutable objects (like arrays of size 1—unless the state object itself is mutable already).